### ISSUE TITLE
Add support for application addresses with whitelisting

### ISSUE THREAD

**Issue by kantai** [blockstack/gaia#96]
This has to do with governing which addresses a gaia hub will accept writes on behalf of, which is a necessity for operating private hubs (https://github.com/blockstack/gaia#private-hubs)

Currently, the `config.json` file governs address whitelisting in Gaia. If a user wishes to restrict the set of addresses which can write to the gaia hub, this can pose a major usability issue.

Basically, if you want to whitelist an "identity address" like `aaron.id`'s `1EJh2y3xKUwFjJ8v29a2NRruPJ71neozEE` -- you cannot _actually_ derive, say, my application address for app `foo` from the ID (or even the `xpub`, because these are all bip32 hardened indexes)

There's a couple of approaches we could take:

1. Add endpoint to the gaia hub which will allow the blockstack-browser to modify the whitelist. This would require the notion of "admin addresses" which can modify the whitelist (you wouldn't want applications, for example, to have this access)

2. `blockstack.js` attaches a "proof" in the authentication header on gaia hub writes. This proof would show that the app address is derived from the identity address. The auth response token _already_ proves this fact (by signing the JWT with the identity key, and including the app private key) --- however, this would require (1) sharing the auth response token and (2) the app's transit key, which we should _not_ do. Instead, we'd probably want to include this as a new field in the auth response as a JWT with `payload = {  appPublicKey, identityPublicKey }` and `signer = identityPublicKey`

I'm in favor of (2), as that's probably the simplest to implement, and it keeps the gaia hub basically stateless (you would just whitelist _identity addresses_, and then the application addresses for those identities would follow from there.)

Would love to hear thoughts on this -- tagging some people who I remember having views on this: @jcnelson @jackzampolin @larrysalibra

**Comment by jackzampolin** [comment]
I agree that 2 would be far and away easiest to implement, however I think an endpoint to manage the configuration would be a really nice feature that we would likely end up building out in the future.

This feature would be a nice enabler for hosted gaia hubs (the service would want to automatically adjust the whitelist based on an outside DB)

**Comment by kantai** [comment]
> This feature would be a nice enabler for hosted gaia hubs (the service would want to automatically adjust the whitelist based on an outside DB)

Outside DBs for the Gaia hub are exactly the kind of thing I think we should try really *really* hard to avoid -- because this basically means that we now have this additional piece of user-state which is neither really gaia-storage data, nor data that lives on-chain, and then that means we'll need additional processes around migration and recovery there.

I think an example makes this a little more clear: let's say your gaia hub loses this configured data (for whatever reason). Now, all of your applications will need to re-auth with the `add-address-to-allowed-set` endpoint. Apps can't do that on their own-- the blockstack browser will need to do that-- but the browser doesn't know all of the applications you've logged into (it doesn't track that, and you might have reset your browser).

**Comment by jcnelson** [comment]
> blockstack.js attaches a "proof" in the authentication header on gaia hub writes. This proof would show that the app address is derived from the identity address. The auth response token already proves this fact (by signing the JWT with the identity key, and including the app private key) --- however, this would require (1) sharing the auth response token and (2) the app's transit key, which we should not do. Instead, we'd probably want to include this as a new field in the auth response as a JWT with payload = { appPublicKey, identityPublicKey } and signer = identityPublicKey

I have a couple of questions.

First, the application is signing the "proof" JWT with the app private key, right?  I'm not clear on how the "proof" is generated by blockstack.js.  I'm not sure why `signer = identityPublicKey` here.

Second, regardless of how it is generated, could the "proof" be used by a malicious actor to replay a write operation?  I'd recommend that the "proof" should have a monotonically-increasing timestamp and a short-lived lifetime to prevent reuse (e.g. no more than 30 seconds).  The Gaia hub should NACK a request if its timestamp is lower than the last-seen timestamp (or is over 30 seconds old).

**Comment by kantai** [comment]
> First, the application is signing the "proof" JWT with the app private key, right? I'm not clear on how the "proof" is generated by blockstack.js. I'm not sure why signer = identityPublicKey here.

The "proof" is a signed assertion that a given public-key-hash is a child of the signer's pub-key-hash, which should be the _identity_ address.

 For example, an assertion that `1Jrfd4ZHTd7ks5PdRHzBvmehZZVFvUFobR` is a child of  `16LPutBwZQXy9BypANh1q2y5Z5fpwrmifp` (which you can find out from my profile.json-- https://core.blockstack.org/v1/users/aaron.blockstack_berlin.id) -- you need to be able to do this if you want to run a gaia hub which only authenticates requests from "16LPutBwZQXy9BypANh1q2y5Z5fpwrmifp" and _also_ it's child-keys.

> Second,  regardless of how it is generated, could the "proof" be used by a  malicious actor to replay a write operation?  I'd recommend that the  "proof" should have a monotonically-increasing timestamp and a  short-lived lifetime to prevent reuse (e.g. no more than 30 seconds).   The Gaia hub should NACK a request if its timestamp is lower than the last-seen timestamp (or is over 30 seconds old).

We currently use _really_ long lived tokens -- I agree this is a problem, and should be addressed, but I think that's somewhat separate -- here I'm not trying to authenticate, I'm just trying to get the gaia hub to associate an _app address_ with a given identity address.

**Comment by jcnelson** [comment]
The signer would be the identity private key, not the public key, right?  The proof would have to be generated in the browser (e.g. on handling a sign-in), and passed to the application as a bearer token?

**Comment by kantai** [comment]
> The signer would be the identity private key, not the public key, right? The proof would have to be generated in the browser (e.g. on handling a sign-in), and passed to the application as a bearer token?

Yep, exactly.

**Comment by larrysalibra** [comment]
+1 to option 2. Keep additional state out of gaia.

**Comment by jackzampolin** [comment]
I have been thinking about services writing to gaia on behalf of applications and have been thinking of a usecase like API integrations to apps. One example would be generation of an index of your blocktweet friend's tweets (timeline) or a Salesforce integration with Graphite Tables. These could easily be modular little programs with scripted (heroku?) deploys.

**Comment by kantai** [comment]
Here's a proposed spec for the authentication:

The payload used by the jsontoken supplied by the child address would look like:

```javascript
{
  gaiaChallenge: challengeText,
  iss: appPublicKey,
  exp: expirationTime,
  salt: '...',
  associationToken: {
    header: { typ: 'JWT', alg: 'ES256K' },
    signature: '...',
    payload: {
      iss: identityPublicKey,
      exp: expirationTimeOfAssociation,
      childToAssociate: appPublicKey,
      salt
    }
  }
}
```

So, that's a jsontoken in a json token.

**Comment by jcnelson** [comment]
PR sent
